front/src/applications/operationalStudies/components/MacroEditor/OsrdNgeSync.tsx

```diff
@@ -0,0 +1,918 @@
+            macroNodeForm: node,
+          }
+        )
+      );
+      const newNode = await createPromise.unwrap();
+      this.indexNode(newNode, true);
+    } catch (e) {
+      console.error(e);
+    }
+  }
+
+  private async apiUpdateNode(node: MacroNodeResponse) {
+    try {
+      await this.dispatch(
+        osrdEditoastApi.endpoints.putProjectsByProjectIdStudiesAndStudyIdScenariosScenarioIdMacroNodesNodeId.initiate(
+          {
+            projectId: this.scenario.project.id,
+            studyId: this.scenario.study_id,
+            scenarioId: this.scenario.id,
+            nodeId: node.id,
+            macroNodeForm: node,
+          }
+        )
+      );
+      this.indexNode(node, true);
+    } catch (e) {
+      console.error(e);
+    }
+  }
+
+  private async apiDeleteNode(node: MacroNodeResponse) {
+    try {
+      await this.dispatch(
+        osrdEditoastApi.endpoints.deleteProjectsByProjectIdStudiesAndStudyIdScenariosScenarioIdMacroNodesNodeId.initiate(
+          {
+            projectId: this.scenario.project.id,
+            studyId: this.scenario.study_id,
+            scenarioId: this.scenario.id,
+            nodeId: node.id,
+          }
+        )
+      );
+      this.deleteNodeByKey(node.path_item_key);
+    } catch (e) {
+      console.error(e);
+    }
+  }
+
+  /**
+   * Cast a NGE node to a node.
+   */
+  static castNgeNode(
+    node: NetzgrafikDto['nodes'][0],
+    labels: NetzgrafikDto['labels']
+  ): Omit<MacroNodeResponse, 'path_item_key'> {
+    return {
+      id: node.id,
+      trigram: node.betriebspunktName,
+      full_name: node.fullName,
+      connection_time: node.connectionTime,
+      position_x: node.positionX,
+      position_y: node.positionY,
+      labels: node.labelIds
+        .map((id) => {
+          const ngeLabel = labels.find((e) => e.id === id);
+          if (ngeLabel) return ngeLabel.label;
+          return null;
+        })
+        .filter((n) => n !== null) as string[],
+    };
+  }
+
+  /**
+   * Given an path step, returns its opId
+   */
+  static getPathKey(item: TrainScheduleResult['path'][0]): string {
+    if ('trigram' in item)
+      return `trigram:${item.trigram}${item.secondary_code ? `/${item.secondary_code}` : ''}`;
+    if ('operational_point' in item) return `op_id:${item.operational_point}`;
+    if ('uic' in item) return `uic:${item.uic}`;
+    if ('track' in item && 'offset' in item) return `track_offset:${item.track}+${item.offset}`;
+
+    // Error
+    console.error(`Can't find an path ID for ${JSON.stringify(item)}`);
+    throw new Error(`Can't find an path ID for ${JSON.stringify(item)}`);
+  }
+
+  /**
+   * Given a search result item, returns all possible opIds, ordered by weight.
+   */
+  static getPathKeys(item: SearchResultItemOperationalPoint): string[] {
+    const result = [];
+    result.push(`op_id:${item.obj_id}`);
+    result.push(`trigram:${item.trigram}${'ch' in item ? `/${item.ch}` : ''}`);
+    result.push(`uic:${item.uic}`);
```

Comment:
Ditto

Reply:
I can't see `/ch` in the updated PR - has this been perhaps been lost in a rebase or something?

Reply:
Ah - it seems like the refacto still has this issue, and then the last commit fixes it.